File: Competitive Programming 3/Section 1.3.3/119 - Greedy Gift Givers/Python/main.py
from sys import stdin

def main():
    noFirst = False
    while True:
        try:
            n = int(input())
        except:
            break
        if noFirst:
            print()
        noFirst=True
        gainLoss=[]
        toPosition={}
        #register people
        personNames=input().split()
        for i,personName in enumerate(personNames):
            gainLoss.append(0)
            toPosition[personName]=i
        #simulate
        for _ in range(0,n):
            tk = input().split()
            person,money,numberOfGifts = tk[0],int(tk[1]),int(tk[2])
            evenMoney = money//numberOfGifts if numberOfGifts!=0 else 0
            position = toPosition[person]
            gainLoss[position]+=-evenMoney*numberOfGifts
            for j in range(0,numberOfGifts):
                gainLoss[toPosition[tk[3+j]]]+=evenMoney
        for personName,gainLossI in zip(personNames,gainLoss):
            print(str(personName)+" "+str(gainLossI))
main()

File: Competitive Programming 3/Section 1.3.3/119 - Greedy Gift Givers/C++/main.cpp
#include<bits/stdc++.h>

using namespace std;

int gainLoss[12];
string personName[12];
int n;

int main(){
    ios_base::sync_with_stdio(0);cin.tie(0);
    for(bool noFirst=false;cin>>n;noFirst=true){
        if(noFirst){
            cout << "\n";
        }
        map<string,int>toPosition;
        //register people
        for(int i = 0 ; i < n ; i ++){
            cin >> personName[i];
            gainLoss[i]=0;
            toPosition.insert(make_pair(personName[i],i));
        }
        //simulate
        for(int i = 0 ; i < n ; i ++){
            string person;
            cin >> person;
            int money,numberOfGifts,evenMoney;
            cin >> money>>numberOfGifts;
            evenMoney = numberOfGifts?money/numberOfGifts:0;
            gainLoss[toPosition[person]]+=-evenMoney*numberOfGifts;
            for(int j = 0 ; j < numberOfGifts; j ++){
                string toPerson;
                cin >> toPerson;
                gainLoss[toPosition[toPerson]]+=evenMoney;
            }
        }
        for(int i = 0 ; i < n ; i ++){
            cout << personName[i] << " " << gainLoss[i] << "\n";
        }
    }
    return 0;
}